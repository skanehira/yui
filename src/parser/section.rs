use crate::{
    elf::section::{Header, SectionFlag, SectionType},
    parser::error::ParseError,
};
use nom::{
    Parser as _,
    combinator::map_res,
    multi::count,
    number::complete::{le_u32, le_u64},
};

use super::{ParseResult, helper};

impl TryFrom<u32> for SectionType {
    type Error = ParseError;
    fn try_from(value: u32) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Null),
            1 => Ok(Self::ProgBits),
            2 => Ok(Self::SymTab),
            3 => Ok(Self::StrTab),
            4 => Ok(Self::Rela),
            5 => Ok(Self::Hash),
            6 => Ok(Self::Dynamic),
            7 => Ok(Self::Note),
            8 => Ok(Self::NoBits),
            9 => Ok(Self::Rel),
            10 => Ok(Self::ShLib),
            11 => Ok(Self::DynSym),
            14 => Ok(Self::InitArray),
            15 => Ok(Self::FiniArray),
            16 => Ok(Self::PreInitArray),
            17 => Ok(Self::Group),
            18 => Ok(Self::SymTabShndx),
            0x6ffffffa => Ok(Self::GnuVerdef),
            0x6ffffffb => Ok(Self::GnuVerneed),
            0x6ffffffc => Ok(Self::GnuVersym),
            _ => Err(ParseError::InvalidSectionType(value)),
        }
    }
}

fn parse_type(raw: &[u8]) -> ParseResult<SectionType> {
    map_res(le_u32, SectionType::try_from).parse(raw)
}

fn parse_flags(raw: &[u8]) -> ParseResult<Vec<SectionFlag>> {
    map_res(le_u64, |mask| -> Result<Vec<SectionFlag>, ParseError> {
        let flag_variants = [
            SectionFlag::Write,
            SectionFlag::Alloc,
            SectionFlag::ExecInstr,
            SectionFlag::Merge,
            SectionFlag::Strings,
            SectionFlag::InfoLink,
            SectionFlag::LinkOrder,
            SectionFlag::Group,
            SectionFlag::Tls,
            SectionFlag::Compressed,
        ];

        if mask == 0 {
            return Ok(vec![]);
        }

        let flags = flag_variants
            .into_iter()
            .filter(|flag| mask & (*flag as u64) != 0)
            .collect();

        Ok(flags)
    })
    .parse(raw)
}

/// Parses the section headers from the raw ELF file data.
///
/// # Arguments
///
/// * `raw` - A byte slice containing the raw ELF file data.
/// * `shoff` - The offset in the file where the section header table begins.
/// * `shstrndx` - The index of the section header string table in the section header table.
/// * `shnum` - The number of section headers in the section header table.
///
/// # Returns
///
/// A `ParseResult` containing a vector of parsed `Header` structures or a `ParseError` if parsing fails.
///
/// # Notes
///
/// If `shnum` is 0, the function returns an empty vector of headers. The function uses a parser combinator
/// to iterate through the section headers and extract their fields. After parsing, it resolves the section
/// names using the section header string table.
pub fn parse_header(
    raw: &[u8],
    shoff: usize,
    shstrndx: usize,
    shnum: usize,
) -> ParseResult<Vec<Header>> {
    if shnum == 0 {
        return Ok((raw, vec![]));
    }

    count(
        |rest| {
            let (rest, name_idx) = le_u32(rest)?;
            let (rest, r#type) = parse_type(rest)?;
            let (rest, flags) = parse_flags(rest)?;
            let (rest, addr) = le_u64(rest)?;
            let (rest, offset) = le_u64(rest)?;
            let (rest, size) = le_u64(rest)?;
            let (rest, link) = le_u32(rest)?;
            let (rest, info) = le_u32(rest)?;
            let (rest, addralign) = le_u64(rest)?;
            let (rest, entsize) = le_u64(rest)?;
            let data = raw[offset as usize..(offset + size) as usize].to_vec();

            let header = Header {
                name_idx,
                name: "".into(), // to be filled later
                r#type,
                flags,
                addr,
                offset,
                size,
                link,
                info,
                addralign,
                entsize,
                section_raw_data: data,
            };

            Ok((rest, header))
        },
        shnum,
    )
    .parse(&raw[shoff..])
    .map(|(rest, mut headers)| {
        let string_table = &raw[headers[shstrndx].offset as usize..];
        for header in headers.iter_mut() {
            header.name = helper::get_string_by_offset(string_table, header.name_idx as usize);
        }
        (rest, headers)
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn should_parse_section_header_table() {
        let raw = include_bytes!("./fixtures/sub.o");
        let (_, header) = crate::parser::header::parse(raw).unwrap();

        let (_, header_table) = parse_header(
            raw,
            header.shoff as usize,
            header.shstrndx as usize,
            header.shnum as usize,
        )
        .unwrap();

        insta::assert_debug_snapshot!(header_table);
    }
}
