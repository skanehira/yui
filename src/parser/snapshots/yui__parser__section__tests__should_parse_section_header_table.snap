---
source: src/parser/section.rs
expression: header_table
---
[
    Header {
        name_idx: 0,
        name: "",
        type: Null,
        flags: [],
        addr: 0,
        offset: 0,
        size: 0,
        link: 0,
        info: 0,
        addralign: 0,
        entsize: 0,
        section_raw_data: [],
    },
    Header {
        name_idx: 27,
        name: ".text",
        type: ProgBits,
        flags: [
            Alloc,
            ExecInstr,
        ],
        addr: 0,
        offset: 64,
        size: 0,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 0,
        section_raw_data: [],
    },
    Header {
        name_idx: 33,
        name: ".data",
        type: ProgBits,
        flags: [
            Write,
            Alloc,
        ],
        addr: 0,
        offset: 64,
        size: 4,
        link: 0,
        info: 0,
        addralign: 4,
        entsize: 0,
        section_raw_data: [
            23,
            0,
            0,
            0,
        ],
    },
    Header {
        name_idx: 39,
        name: ".bss",
        type: NoBits,
        flags: [
            Write,
            Alloc,
        ],
        addr: 0,
        offset: 68,
        size: 0,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 0,
        section_raw_data: [],
    },
    Header {
        name_idx: 44,
        name: ".comment",
        type: ProgBits,
        flags: [
            Merge,
            Strings,
        ],
        addr: 0,
        offset: 68,
        size: 40,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 1,
        section_raw_data: [
            0,
            71,
            67,
            67,
            58,
            32,
            40,
            68,
            101,
            98,
            105,
            97,
            110,
            32,
            49,
            48,
            46,
            50,
            46,
            49,
            45,
            54,
            41,
            32,
            49,
            48,
            46,
            50,
            46,
            49,
            32,
            50,
            48,
            50,
            49,
            48,
            49,
            49,
            48,
            0,
        ],
    },
    Header {
        name_idx: 53,
        name: ".note.GNU-stack",
        type: ProgBits,
        flags: [],
        addr: 0,
        offset: 108,
        size: 0,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 0,
        section_raw_data: [],
    },
    Header {
        name_idx: 1,
        name: ".symtab",
        type: SymTab,
        flags: [],
        addr: 0,
        offset: 112,
        size: 216,
        link: 7,
        info: 8,
        addralign: 8,
        entsize: 24,
        section_raw_data: [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            1,
            0,
            0,
            0,
            4,
            0,
            241,
            255,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            3,
            0,
            1,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            3,
            0,
            2,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            3,
            0,
            3,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            7,
            0,
            0,
            0,
            0,
            0,
            2,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            3,
            0,
            5,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            3,
            0,
            4,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            10,
            0,
            0,
            0,
            17,
            0,
            2,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            4,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
        ],
    },
    Header {
        name_idx: 9,
        name: ".strtab",
        type: StrTab,
        flags: [],
        addr: 0,
        offset: 328,
        size: 12,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 0,
        section_raw_data: [
            0,
            115,
            117,
            98,
            46,
            99,
            0,
            36,
            100,
            0,
            120,
            0,
        ],
    },
    Header {
        name_idx: 17,
        name: ".shstrtab",
        type: StrTab,
        flags: [],
        addr: 0,
        offset: 340,
        size: 69,
        link: 0,
        info: 0,
        addralign: 1,
        entsize: 0,
        section_raw_data: [
            0,
            46,
            115,
            121,
            109,
            116,
            97,
            98,
            0,
            46,
            115,
            116,
            114,
            116,
            97,
            98,
            0,
            46,
            115,
            104,
            115,
            116,
            114,
            116,
            97,
            98,
            0,
            46,
            116,
            101,
            120,
            116,
            0,
            46,
            100,
            97,
            116,
            97,
            0,
            46,
            98,
            115,
            115,
            0,
            46,
            99,
            111,
            109,
            109,
            101,
            110,
            116,
            0,
            46,
            110,
            111,
            116,
            101,
            46,
            71,
            78,
            85,
            45,
            115,
            116,
            97,
            99,
            107,
            0,
        ],
    },
]
